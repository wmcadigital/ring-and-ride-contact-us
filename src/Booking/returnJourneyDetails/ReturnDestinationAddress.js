import { useEffect } from "react";
import PropTypes from "prop-types";
import { useFormState } from "react-final-form";

import FormSection from "../../common/FormSection";
import RadioGroup from "../../common/RadioGroup";
import Question from "../../common/Question";
import RadioButton from "../../common/RadioButton";
import FieldError from "../../common/FieldError";
import ProgressIndicator from "../../common/ProgressIndicator";
import ContactDetails from "../../common/ContactDetails";
import AddressFormatted from "../../common/AddressFormatted";
import getSectionPositionInfo from "../getSectionPosition";
import getAboutReturnJourneySectionName from "./getAboutReturnJourneySectionName";

import { required } from "../../common/validation";

const ReturnDestinationAddress = ({
  setReturnDestinationAddress,
  setGoToPage,
}) => {
  const stateApi = useFormState();
  const formValues = stateApi.values;
  const bookingParty = formValues["bookingParty"];

  const error = stateApi.submitFailed
    ? stateApi.errors?.returnDestinationAddress
    : null;

  let question = "";
  if (bookingParty === "mySelf") {
    question = "What address do you want to go to?";
  } else if (bookingParty === "behalfSomeone") {
    question = `What address does ${formValues["firstName"]} want to go to?`;
  } else if (bookingParty === "behalfGroup") {
    question = "What address is the group going to?";
  }

  let labelOne = "";
  if (bookingParty === "mySelf") {
    labelOne = "My registered home address";
  } else if (bookingParty === "behalfSomeone") {
    labelOne = "Their registered home address";
  } else if (bookingParty === "behalfGroup") {
    labelOne = "The outward journey address they were collected from";
  }

  useEffect(() => {
    if (formValues.returnDestinationAddress) {
      setReturnDestinationAddress(formValues.returnDestinationAddress);
    }
    if (formValues.returnDestinationAddress === "other") {
      setGoToPage(null);
    }
  }, [formValues.returnDestinationAddress]);

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition={getSectionPositionInfo("3", stateApi)}
        sectionName={getAboutReturnJourneySectionName(stateApi)}
      />
      <Question text={question} />
      <RadioGroup error={error}>
        <FieldError text={error} />
        <RadioButton
          key={1}
          label={labelOne}
          validation={required}
          value="sameAsOutwardCollectionAddress"
          fieldName="returnDestinationAddress"
        />
        {formValues.returnDestinationAddress ===
          "sameAsOutwardCollectionAddress" && bookingParty === "behalfGroup" ? (
          <div className="wmrards-m-l-xl wmrards-p-l-sm wmrards-m-b-lg">
            <ContactDetails>
              <AddressFormatted addressObj={formValues["otherOutward"]} />
            </ContactDetails>
          </div>
        ) : null}
        <RadioButton
          key={2}
          label="Another address"
          validation={required}
          value="other"
          fieldName="returnDestinationAddress"
        />
      </RadioGroup>
    </FormSection>
  );
};

ReturnDestinationAddress.propTypes = {
  setReturnDestinationAddress: PropTypes.func,
  setGoToPage: PropTypes.func,
};

ReturnDestinationAddress.defaultProps = {
  setReturnDestinationAddress: () => {},
  setGoToPage: () => {},
};

export default ReturnDestinationAddress;
