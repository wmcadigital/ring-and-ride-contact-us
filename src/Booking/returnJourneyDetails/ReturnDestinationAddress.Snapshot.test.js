import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import ReturnDestinationAddress from "./ReturnDestinationAddress";

jest.mock("react-final-form");

describe("ReturnDestinationAddress - snapshots", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<ReturnDestinationAddress />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "behalfSomeone",
        firstName: "John",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<ReturnDestinationAddress />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for a group", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "behalfGroup",
        firstName: "John",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<ReturnDestinationAddress />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
        firstName: "John",
      },
      submitFailed: true,
      errors: { returnDestinationAddress: "Required" },
    }));
    const renderer = createRenderer();

    renderer.render(<ReturnDestinationAddress />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
