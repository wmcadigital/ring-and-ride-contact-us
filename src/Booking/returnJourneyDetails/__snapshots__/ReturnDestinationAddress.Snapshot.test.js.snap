// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReturnDestinationAddress - snapshots individual is filling out form for a group 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan their return journey"
    sectionPosition="Section 4 of 5"
  />
  <Question
    text="What address is the group going to?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="returnDestinationAddress"
      label="The outward journey address they were collected from"
      onClickCallback={[Function]}
      validation={[Function]}
      value="sameAsOutwardCollectionAddress"
    />
    <RadioButton
      fieldName="returnDestinationAddress"
      label="Another address"
      onClickCallback={[Function]}
      validation={[Function]}
      value="other"
    />
  </RadioGroup>
</FormSection>
`;

exports[`ReturnDestinationAddress - snapshots individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan their return journey"
    sectionPosition="Section 3 of 4"
  />
  <Question
    text="What address does John want to go to?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="returnDestinationAddress"
      label="Their registered home address"
      onClickCallback={[Function]}
      validation={[Function]}
      value="sameAsOutwardCollectionAddress"
    />
    <RadioButton
      fieldName="returnDestinationAddress"
      label="Another address"
      onClickCallback={[Function]}
      validation={[Function]}
      value="other"
    />
  </RadioGroup>
</FormSection>
`;

exports[`ReturnDestinationAddress - snapshots individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan your return journey"
    sectionPosition="Section 3 of 4"
  />
  <Question
    text="What address do you want to go to?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="returnDestinationAddress"
      label="My registered home address"
      onClickCallback={[Function]}
      validation={[Function]}
      value="sameAsOutwardCollectionAddress"
    />
    <RadioButton
      fieldName="returnDestinationAddress"
      label="Another address"
      onClickCallback={[Function]}
      validation={[Function]}
      value="other"
    />
  </RadioGroup>
</FormSection>
`;

exports[`ReturnDestinationAddress - snapshots renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan your return journey"
    sectionPosition="Section 3 of 4"
  />
  <Question
    text="What address do you want to go to?"
  />
  <RadioGroup
    error="Required"
  >
    <FieldError
      text="Required"
    />
    <RadioButton
      fieldName="returnDestinationAddress"
      label="My registered home address"
      onClickCallback={[Function]}
      validation={[Function]}
      value="sameAsOutwardCollectionAddress"
    />
    <RadioButton
      fieldName="returnDestinationAddress"
      label="Another address"
      onClickCallback={[Function]}
      validation={[Function]}
      value="other"
    />
  </RadioGroup>
</FormSection>
`;
